lib/hanami/slice.rb

```diff
@@ -113,12 +113,69 @@ def define_container
           config.root = root
           config.bootable_dirs = ["config/boot"]
 
+          # Add the "lib" component dir; all slices will load components from lib
           if root.join("lib").directory?
-            config.component_dirs.add "lib" do |dir|
-              dir.default_namespace = namespace_path.tr(File::SEPARATOR, config.namespace_separator)
+            config.component_dirs.add("lib") do |component_dir|
+              if application.configuration.autoloader
+                # When using an autoloader, expect component files in the root of the lib
+                # component dir to define classes inside the slice's namespace.
+                #
+                # e.g. "lib/foo.rb" should define SliceNamespace::Foo, and will be
+                # registered as "foo"
+                component_dir.namespaces.root(key: nil, const: namespace_path)
+
+                application.configuration.autoloader.push_dir(root.join("lib"), namespace: namespace)
+              else
+                # When not using an autoloader, expect component files in lib to follow
+                # Ruby conventions, in which their file path matches their class constant.
+                #
+                # Following from this, add the slice's namespace as a configured component
+                # dir namespace, so that the slice's namespace is not repeated in every
+                # container key.
+                #
+                # e.g.
+                #
+                # - "lib/[slice_namespace]/foo.rb" (i.e. within the configured component
+                #   dir namespace) should define SliceNamespace::Foo, and will be
+                #   registered as "foo"
+                # - "lib/bar/baz.rb" (i.e. outside the configured component dir namespace)
+                #   should define Bar::Baz, and will be registered as "bar.baz"
+                component_dir.namespaces.add(namespace_path, key: nil)
+              end
             end
+          end
+
+          # When using an autoloader, add additional component dirs for each if the
+          # configured component dir paths (if they exist)
+          if application.configuration.autoloader
+            application.configuration.component_dir_paths.each do |slice_dir|
```

Comment:
> By "conflicting names" you mean file names, right? And so the idea of using $LOAD_PATH doesn't work because require "post" is ambiguous and may lead to unexpected results.
>
> Just double checking this with you, that we're talking about path name and not class name.
Because I expect these entities to have different namespaces: e.g. Admin::Entities::Post.

Yep, both of those statements are exactly right :)

Your idea about adding the slice root to the load path is an interesting one!

So to play this out fully, given these files:

```
slices/admin/entities/post.rb    # Admin::Entities::Post
slices/admin/lib/create_post.rb  # Admin::CreatePost

slices/main/entities/post.rb     # Main::Entities::Post
slices/main/lib/create_post.rb   # Main::CreatePost
```

We _would_ be able to properly require those entity files if we added the actual slices dir (i.e. `[app_root]/slices` itself, _not_ each individual slice dir) to the load path. Because in that case, `require "admin/entities/post" would properly maps onto the real filesystem path for the `slices/admin/entities/post.rb` file.

So while that works for the entities, there's still one last problem with this arrangement, in that there's still no way to require anything inside the `slices/[slice_name]/lib` directories, which is where we expect most of the per-slice business logic to be contained. For those files, we're stuck with the same problem I described initially, because:

1. There's no way for `require "admin/create_post"` to work (because `admin/create_post` does not map directly to any real filesystem path), and
2. For the same reasons as I originally described, we still can't add each slice's `lib/` directly onto `$LOAD_PATH`, because we'd have that same problem where it would then need `require "create_post"`, but that then becomes a require that will only work for whichever slice's `lib/` happens to be first on the `$LOAD_PATH`.

So this leads me back to those two options I shared before:

> 1. Make it so autoloading is always on, with no opt out
> 2. Keep the opt out in place, but accept and make it clear that if a user does opt out, they need to maintain a completely set of file hierarchies in their source directories

And my sense for the best option of these two is the same, too:

> Given that option (2) makes generators and documentation and many other things much more complicated, and given that I expect autoloading would be disabled only by the tiniest fraction of our users, I propose that we make it so that autoloading is always on.

What do you reckon?

Reply:
@timriley I'm fine if they will be able to turn off autoloading, and need to manually do: `require "admin/lib/create_post"`.

We're talking about a tiny fraction of people who:

- Will disable autoloading
- Will put code outside of the canonical directories

If that require above looks wrong for them, they can still create an `admin/operations` directory, so the require will look better: `require "admin/operations/create_post"`.

---

To recap, I would:

- Give the ability to turn off autoloading
- Add `[root]/slices` to `$LOAD_PATH` to have nice manual requires from canonical directories. Example: `require "admin/entities/post"`.
- Ignore other ugly manual requires. Example: `require "admin/lib/create_post"`